security/src/main/java/io/micronaut/security/authentication/Authenticator.java

```diff
@@ -45,55 +62,156 @@ public class Authenticator<T> {
 
     private static final Logger LOG = LoggerFactory.getLogger(Authenticator.class);
 
-    protected final Collection<AuthenticationProvider<T>> authenticationProviders;
+    /**
+     *
+     * @deprecated Unused. To be removed in the next major version.
+     */
+    @Deprecated(forRemoval = true, since = "4.5.0")
+    protected final Collection<io.micronaut.security.authentication.AuthenticationProvider<T>> authenticationProviders;
+
+    private final List<ReactiveAuthenticationProvider<T>> reactiveAuthenticationProviders;
+
+    private final BeanContext beanContext;
+    private final List<AuthenticationProvider<T>> imperativeAuthenticationProviders;
     private final SecurityConfiguration securityConfiguration;
 
+    private final Scheduler scheduler;
+
     /**
-     * @param authenticationProviders A list of available authentication providers
+     * @param beanContext Bean Context
+     * @param reactiveAuthenticationProviders A list of available Reactive authentication providers
+     * @param authenticationProviders A list of available imperative authentication providers
+     * @param executorService BLOCKING executor service
      * @param securityConfiguration The security configuration
      */
-    public Authenticator(Collection<AuthenticationProvider<T>> authenticationProviders,
+    @Inject
+    public Authenticator(BeanContext beanContext,
+                         List<ReactiveAuthenticationProvider<T>> reactiveAuthenticationProviders,
```

Comment:
since you are injecting lists here annotate the `ReactiveAuthenticationProvider` and `AuthenticationProvider` interfaces with `@Indexed` to improvement performance

Reply:
[I annotated both with `@Indexed`](https://github.com/micronaut-projects/micronaut-security/pull/1526/commits/c737dd155138cd26f01ed3e9ab2fb3b9a1d4acfd)